File: src/snippets/rate-limiting/reference/claude-managed-agents/py/TokenBucketRequest.py
import os

from arcjet.guard import TokenBucket, launch_arcjet
from arcjet.guard.claude_managed_agents import guard_custom_tool

arcjet = launch_arcjet(key=os.environ["ARCJET_KEY"])
lookup_limit = TokenBucket(
    refill_rate=5,
    interval_seconds=10,
    max_tokens=10,
    bucket="lookups",
)


async def lookup_order(arguments: dict) -> dict:
    return {"order_id": arguments["order_id"], "status": "shipped"}


guarded_lookup = guard_custom_tool(
    guard=arcjet,
    tool=lookup_order,
    action="order.looked-up",
    # Deduct 50 tokens from the bucket. The value for `requested` must be a
    # positive integer.
    rules=lambda arguments: [
        lookup_limit(key=arguments["order_id"], requested=50)
    ],
)

File: src/snippets/rate-limiting/reference/claude-managed-agents/js/TokenBucketRequest.ts
import { launchArcjet, tokenBucket } from "@arcjet/guard";
import { guardCustomTool } from "@arcjet/guard/claude-managed-agents/v0";

const arcjet = launchArcjet({ key: process.env.ARCJET_KEY! });

const lookupLimit = tokenBucket({
  bucket: "lookups",
  refillRate: 5,
  intervalSeconds: 10,
  maxTokens: 10,
});

export const lookupOrder = guardCustomTool(
  arcjet,
  async ({ orderId }: { orderId: string }) => ({
    orderId,
    status: "shipped",
  }),
  {
    action: "order.looked-up",
    rules: ({ orderId }) => [
      // Deduct 50 tokens from the bucket.
      // The value for `requested` must be a positive integer.
      lookupLimit({ key: orderId, requested: 50 }),
    ],
  },
);
